src/Elf.cpp

```diff
@@ -0,0 +1,989 @@
+    dynamic_sym.set_type(Symbol::STT_OBJECT);
+    got.append_contents((addr_t)0);
+    // On some platforms, GOT slots 1 and 2 are also reserved.
+    got.append_contents((addr_t)0);
+    got.append_contents((addr_t)0);
+
+    // Since we can't change the object, start a map of all of the
+    // symbols that we can mutate. If a symbol from the object is a
+    // key in this map, we use the mapped value instead.
+    std::map<const Symbol *, const Symbol *> symbols;
+    symbols[&dynamic_sym] = &dynamic_sym;
+    for (const Symbol &i : obj.symbols()) {
+        if (i.get_name() == "_GLOBAL_OFFSET_TABLE_") {
+            symbols[&i] = &got_sym;
+        } else {
+            symbols[&i] = &i;
+        }
+    }
+
+    // Get a symbol from a relocation, accounting for the symbol map
+    // above.
+    auto get_symbol = [&](const Relocation &r) {
+        const Symbol *sym = r.get_symbol();
+        if (!sym) {
+            return sym;
+        }
+
+        auto i = symbols.find(sym);
+        if (i != symbols.end()) return i->second;
+        return sym;
+    };
+
+    // Check if a relocation needs a PLT entry, which adds some
+    // additional conditions on top of what the linker implementation
+    // wants.
+    auto needs_plt_entry = [&](const Relocation &r) {
+        const Symbol *s = get_symbol(r);
+        if (!s || s->is_defined()) {
+            return false;
+        }
+
+        if (s->get_type() != Symbol::STT_NOTYPE) {
+            return false;
+        }
+
+        return linker->needs_plt_entry(r);
+    };
+
+    // We need to build the PLT, so it can be positioned along with
+    // the rest of the text sections.
+    Section plt(".plt", Section::SHT_PROGBITS);
+    plt.set_alignment(16);
+    plt.set_flags(Section::SHF_ALLOC | Section::SHF_EXECINSTR);
+    std::list<Symbol> plt_symbols;
+    std::map<const Symbol *, const Symbol *> plt_defs;
+    // Hack: We're defining the global offset table, so it shouldn't be treated as an external symbol.
+    plt_defs[&got_sym] = &got_sym;
+    for (const Section &s : obj.sections()) {
+        for (const Relocation &r : s.relocations()) {
+            if (!needs_plt_entry(r)) {
+                continue;
+            }
+
+            const Symbol *sym = get_symbol(r);
+            const Symbol *& plt_def = plt_defs[sym];
+            if (plt_def) {
+                // We already made a PLT entry for this symbol.
+                continue;
+            }
+
+            debug(2) << "Defining PLT entry for " << sym->get_name() << "\n";
+            plt_symbols.push_back(linker->add_plt_entry(*sym, plt, got, got_sym));
+
+            plt_def = &plt_symbols.back();
+            symbols[plt_def] = plt_def;
+        }
+    }
+
+    // Start placing the sections into the shared object.
+
+    // Leave room for the header, and program headers at the beginning of the file.
+    append_zeros(output, sizeof(ehdr));
+    append_zeros(output, sizeof(phdrs[0])*3);
```

Comment:
I am not sure whats 3 here ? Is it one PT_LOAD for text, PT_LOAD for data and PT_DYNAMIC ?

Is there an assumption, if so, we should list that.